### PR TITLE
Bigquery localisation

### PR DESCRIPTION
**Pull request by adrianbr**

### CODE REVIEW

File: dlt/pipeline/typing.py [hunk 1]
```diff
@@ -48,20 +49,19 @@ def default_dataset(self, new_value: str) -> None:
         self.DEFAULT_DATASET = new_value
 
     @classmethod
-    def from_services_dict(cls, services: StrAny, dataset_prefix: str) -> "GCPPipelineCredentials":
+    def from_services_dict(cls, services: StrAny, dataset_prefix: str, location: str) -> "GCPPipelineCredentials":
```
Comment (rudolfix) [inline dlt/pipeline/typing.py:52]: please add a default for `location: str = "US"` to make it backward compatible with all our examples. this code will be removed in v2 anyway so we can take some shortcuts

File: dlt/pipeline/typing.py [hunk 2]
```diff
@@ -48,20 +49,19 @@ def default_dataset(self, new_value: str) -> None:
...
         assert dataset_prefix is not None
-
-        return cls("bigquery", services["project_id"], dataset_prefix, services["client_email"], services["private_key"])
+        return cls("bigquery", services["project_id"], dataset_prefix, services["client_email"], services["private_key"], location or cls.LOCATION)
 
     @classmethod
-    def from_services_file(cls, services_path: str, dataset_prefix: str) -> "GCPPipelineCredentials":
+    def from_services_file(cls, services_path: str, dataset_prefix: str, location: str) -> "GCPPipelineCredentials":
```
Comment (rudolfix) [inline dlt/pipeline/typing.py:57]: same here

File: dlt/pipeline/typing.py [hunk 3]
```diff
@@ -48,20 +49,19 @@ def default_dataset(self, new_value: str) -> None:
...
         with open(services_path, "r", encoding="utf-8") as f:
             services = json.load(f)
-        return GCPPipelineCredentials.from_services_dict(services, dataset_prefix)
+        return GCPPipelineCredentials.from_services_dict(services, dataset_prefix, location)
 
     @classmethod
-    def default_credentials(cls, dataset_prefix: str, project_id: str = None) -> "GCPPipelineCredentials":
-        return cls("bigquery", project_id, dataset_prefix, None)
+    def default_credentials(cls, dataset_prefix: str, project_id: str = None, location: str = None) -> "GCPPipelineCredentials":
+        return cls("bigquery", project_id, dataset_prefix, None, None, location or cls.LOCATION)
```
Comment (rudolfix) [inline dlt/pipeline/typing.py:64]: cool!

File: dlt/load/bigquery/client.py [hunk 4]
```diff
@@ -62,7 +62,7 @@ def open_connection(self) -> None:
             credentials = None
         else:
             credentials = service_account.Credentials.from_service_account_info(self.C.as_credentials())
-        self._client = bigquery.Client(self.C.PROJECT_ID, credentials=credentials)
+        self._client = bigquery.Client(self.C.PROJECT_ID, credentials=credentials, location=self.C.LOCATION)
```
Comment (rudolfix) [inline dlt/load/bigquery/client.py:69]: can you do an experiment?
line: 204
```
return BigQueryLoadJob(
                JobClientBase.get_file_name_from_file_path(file_path),
                self._create_load_job(table["name"], table["write_disposition"], file_path),
                self.CREDENTIALS
            )
            
```

replace self.CREDENTIALS with `self.sql_client.native_connection()`

and see if all works. then we are sure the same client contexts used in other places is passed to the load job

File: dlt/pipeline/typing.py (hunk 1 above)
Comment (rudolfix) [reply dlt/pipeline/typing.py:52]: see the linter output
```
./check-package.sh
poetry run mypy --config-file mypy.ini dlt examples
examples/sync_schema_example.py:4: error: Missing positional argument "location" in call to "from_services_file" of "GCPPipelineCredentials"
examples/singer_tap_jsonl_example.py:12: error: Missing positional argument "location" in call to "from_services_file" of "GCPPipelineCredentials"
examples/singer_tap_example.py:10: error: Missing positional argument "location" in call to "from_services_file" of "GCPPipelineCredentials"
examples/quickstart.py:25: error: Missing positional argument "location" in call to "from_services_dict" of "GCPPipelineCredentials"
examples/google_sheets.py:9: error: Missing positional argument "location" in call to "from_services_file" of "GCPPipelineCredentials"
examples/google_drive_csv.py:6[8](https://github.com/scale-vector/dlt/actions/runs/3052952945/jobs/4922951182#step:8:9): error: Missing positional argument "location" in call to "from_services_file" of "GCPPipelineCredentials"
examples/demo_example.py:84: error: Missing positional argument "location" in call to "from_services_file" of "GCPPipelineCredentials"
examples/rasa_example.py:3[9](https://github.com/scale-vector/dlt/actions/runs/3052952945/jobs/4922951182#step:8:10): error: Missing positional argument "location" in call to "from_services_dict" of "GCPPipelineCredentials"
Found 8 errors in 8 files (checked [11](https://github.com/scale-vector/dlt/actions/runs/3052952945/jobs/4922951182#step:8:12)4 source files)
make: *** [Makefile:41: lint] Error 1
Error: Process completed with exit code 2.
```

**Review by rudolfix — CHANGES_REQUESTED** [review]
remember to `make lint` before you push the code

File: dlt/load/bigquery/client.py (hunk 4 above)
Comment (adrianbr) [reply dlt/load/bigquery/client.py:69]: there are 2 places, start_file_load, and restore_file_load, should I just do the restore or both?

File: dlt/load/bigquery/client.py [hunk 5]
```diff
@@ -203,7 +203,8 @@ def restore_file_load(self, file_path: str) -> LoadJob:
             return BigQueryLoadJob(
                 JobClientBase.get_file_name_from_file_path(file_path),
                 self._retrieve_load_job(file_path),
-                self.CREDENTIALS
+                #self.CREDENTIALS
```
Comment (rudolfix) [inline dlt/load/bigquery/client.py:206]: I retract sorry :/

File: tests/load/bigquery/test_bigquery_client.py [hunk 6]
```diff
@@ -86,6 +86,25 @@ def test_bigquery_job_errors(client: BigQueryClient, file_storage: FileStorage)
...
+@pytest.mark.skip()
+def test_bigquery_location(client: BigQueryClient, file_storage: FileStorage) -> None:
+    user_table_name = prepare_event_user_table(client)
+    load_json = {
+        "_dlt_id": uniq_id(),
+        "_dlt_root_id": uniq_id(),
+        "sender_id":'90238094809sajlkjxoiewjhduuiuehd',
+        "timestamp": str(pendulum.now())
+    }
+    #job = expect_load_file(client, file_storage, json.dumps(load_json), user_table_name)
```
Comment (rudolfix) [inline tests/load/bigquery/test_bigquery_client.py:98]: what is the problem with the test? let's set up the testing so you can run tests locally

**Review by rudolfix — CHANGES_REQUESTED** [review]

**Review by rudolfix — APPROVED** [review]

**Pull request merged**